### PR TITLE
[Java] [JDBC] Refactor Tests Failures With Boolean Data Conversion

### PR DESCRIPTION
**Pull request by iurysalino**
Refactor Boolean data Conversion based in this [doc](https://download.oracle.com/otn-pub/jcp/jdbc-4_1-mrel-spec/jdbc4.1-fr-spec.pdf?AuthParam=1646744383_fc21f7eaf2bb05bbcb5736db3c731929) .

Change getBoolean to return a SQLException at arrors cases.

### CODE REVIEW

File: java/flight/flight-jdbc-driver/src/main/java/org/apache/arrow/driver/jdbc/accessor/ArrowFlightJdbcAccessor.java [hunk 1]
```diff
@@ -17,7 +17,7 @@
 
 package org.apache.arrow.driver.jdbc.accessor;
 
-import static org.apache.arrow.driver.jdbc.utils.ExceptionTemplateThrower.getOperationNotSupported;
```
Comment (rafael-telles) [inline java/flight/flight-jdbc-driver/src/main/java/org/apache/arrow/driver/jdbc/accessor/ArrowFlightJdbcAccessor.java:20]: Why this change?

File: java/flight/flight-jdbc-driver/src/main/java/org/apache/arrow/driver/jdbc/accessor/ArrowFlightJdbcAccessorFactory.java [hunk 2]
```diff
@@ -201,7 +201,7 @@ public static ArrowFlightJdbcAccessor createAccessor(ValueVector vector,
       return new ArrowFlightJdbcNullVectorAccessor(setCursorWasNull);
     }
 
-    throw new UnsupportedOperationException();
+    throw new UnsupportedOperationException("This vector is not supported: " + vector.getClass().getName());
```
Comment (rafael-telles) [inline java/flight/flight-jdbc-driver/src/main/java/org/apache/arrow/driver/jdbc/accessor/ArrowFlightJdbcAccessorFactory.java:204]:
```suggestion
    throw new UnsupportedOperationException("Unsupported vector type: " + vector.getClass().getName());
```

File: java/flight/flight-jdbc-driver/src/main/java/org/apache/arrow/driver/jdbc/accessor/impl/text/ArrowFlightJdbcVarCharVectorAccessor.java [hunk 3]
```diff
@@ -100,9 +92,15 @@ public String getString() {
...
-  public boolean getBoolean() {
+  public boolean getBoolean() throws SQLException {
     String value = getString();
-    return value != null && !value.isEmpty() && !value.equals("false") && !value.equals("0");
+    if (value == null || value.equalsIgnoreCase("false") || value.equals("0")) {
+      return false;
+    } else if (value.equalsIgnoreCase("true") || value.equals("1")) {
+      return true;
+    } else {
+      throw new SQLException();
```
Comment (rafael-telles) [inline java/flight/flight-jdbc-driver/src/main/java/org/apache/arrow/driver/jdbc/accessor/impl/text/ArrowFlightJdbcVarCharVectorAccessor.java:102]: Missing message

File: java/flight/flight-jdbc-driver/src/main/java/org/apache/arrow/driver/jdbc/accessor/impl/text/ArrowFlightJdbcVarCharVectorAccessor.java [hunk 4]
```diff
@@ -203,4 +201,12 @@ public Timestamp getTimestamp(Calendar calendar) throws SQLException {
       throw new SQLException(e);
     }
   }
+
+  /**
+   * Functional interface to help integrating VarCharVector and LargeVarCharVector.
+   */
+  @FunctionalInterface
```
Comment (rafael-telles) [inline java/flight/flight-jdbc-driver/src/main/java/org/apache/arrow/driver/jdbc/accessor/impl/text/ArrowFlightJdbcVarCharVectorAccessor.java:208]: Why this change?

File: java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/accessor/impl/text/ArrowFlightJdbcVarCharVectorAccessorTest.java [hunk 5]
```diff
@@ -584,17 +584,21 @@ public void testShouldGetTimestampReturnValidDateWithCalendar() throws Exception
...
+    when(getter.get(0)).thenReturn(value);
+    collector.checkThrows(SQLException.class, () -> accessor.getBoolean());
   }
 
   @Test
+  public void testShouldGetBooleanReturnTrueForNonEmpty() {
+    assertGetBooleanForSQLException(new Text("anything"));
+  }
+
+  @Test(expected = SQLException.class)
```
Comment (rafael-telles) [inline java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/accessor/impl/text/ArrowFlightJdbcVarCharVectorAccessorTest.java:601]: Use `assertGetBooleanForSQLException` for this test too

File: java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/utils/AccessorTestUtils.java [hunk 6]
```diff
@@ -54,30 +67,16 @@ public int getCurrentRow() {
     }
   }
 
-  public interface AccessorSupplier<T extends ArrowFlightJdbcAccessor> {
```
Comment (rafael-telles) [inline java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/utils/AccessorTestUtils.java:57]: Please be careful with unrelated/unnecessary changes like this. This makes reviews harder

**Review by rafael-telles — CHANGES_REQUESTED** [review]

File: java/flight/flight-jdbc-driver/src/main/java/org/apache/arrow/driver/jdbc/accessor/ArrowFlightJdbcAccessor.java (hunk 1 above)
Comment (iurysalino) [reply java/flight/flight-jdbc-driver/src/main/java/org/apache/arrow/driver/jdbc/accessor/ArrowFlightJdbcAccessor.java]: fixed

File: java/flight/flight-jdbc-driver/src/main/java/org/apache/arrow/driver/jdbc/accessor/ArrowFlightJdbcAccessorFactory.java (hunk 2 above)
Comment (iurysalino) [reply java/flight/flight-jdbc-driver/src/main/java/org/apache/arrow/driver/jdbc/accessor/ArrowFlightJdbcAccessorFactory.java]: fixed

File: java/flight/flight-jdbc-driver/src/main/java/org/apache/arrow/driver/jdbc/accessor/impl/text/ArrowFlightJdbcVarCharVectorAccessor.java (hunk 4 above)
Comment (iurysalino) [reply java/flight/flight-jdbc-driver/src/main/java/org/apache/arrow/driver/jdbc/accessor/impl/text/ArrowFlightJdbcVarCharVectorAccessor.java]: fixed

File: java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/accessor/impl/text/ArrowFlightJdbcVarCharVectorAccessorTest.java (hunk 5 above)
Comment (iurysalino) [reply java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/accessor/impl/text/ArrowFlightJdbcVarCharVectorAccessorTest.java]: fixed

File: java/flight/flight-jdbc-driver/src/main/java/org/apache/arrow/driver/jdbc/accessor/impl/text/ArrowFlightJdbcVarCharVectorAccessor.java [hunk 7]
```diff
@@ -100,9 +100,15 @@ public String getString() {
...
-  public boolean getBoolean() {
+  public boolean getBoolean() throws SQLException {
     String value = getString();
-    return value != null && !value.isEmpty() && !value.equals("false") && !value.equals("0");
+    if (value == null || value.equalsIgnoreCase("false") || value.equals("0")) {
+      return false;
+    } else if (value.equalsIgnoreCase("true") || value.equals("1")) {
+      return true;
+    } else {
+      throw new SQLException("Is not possible to convert this value for boolean: " + value);
```
Comment (jcralmeida) [inline java/flight/flight-jdbc-driver/src/main/java/org/apache/arrow/driver/jdbc/accessor/impl/text/ArrowFlightJdbcVarCharVectorAccessor.java:110]: Small change on the message -   "It is not possible..."

File: java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/accessor/impl/numeric/ArrowFlightJdbcBitVectorAccessorTest.java [hunk 8]
```diff
@@ -64,7 +62,8 @@ public void tearDown() {
     this.vectorWithNull.close();
   }
 
-  private <T> void iterate(final Function<ArrowFlightJdbcBitVectorAccessor, T> function,
+  private <T> void iterate(final AccessorTestUtils.AccessorIterator
+      .CheckedFunction<ArrowFlightJdbcBitVectorAccessor, T> function,
```
Comment (jcralmeida) [inline java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/accessor/impl/numeric/ArrowFlightJdbcBitVectorAccessorTest.java:66]: Remove the prefixed words from CheckedFunction. You need to make changes on the import to do this

File: java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/accessor/impl/text/ArrowFlightJdbcVarCharVectorAccessorTest.java [hunk 9]
```diff
@@ -584,19 +584,23 @@ public void testShouldGetTimestampReturnValidDateWithCalendar() throws Exception
...
     when(getter.get(0)).thenReturn(value);
-
     boolean result = accessor.getBoolean();
     collector.checkThat(result, equalTo(expectedResult));
   }
 
+  private void assertGetBooleanForSQLException(Text value) {
+    when(getter.get(0)).thenReturn(value);
+    collector.checkThrows(SQLException.class, () -> accessor.getBoolean());
+  }
```
Comment (jcralmeida) [inline java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/accessor/impl/text/ArrowFlightJdbcVarCharVectorAccessorTest.java:594]: You should do something similiar this in your other PR

Comment (jcralmeida) [reply java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/accessor/impl/text/ArrowFlightJdbcVarCharVectorAccessorTest.java:594]: Also change the name of this method for `assertGetBooleanThrowingSQLException`

**Review by jcralmeida — CHANGES_REQUESTED** [review]

File: java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/accessor/impl/text/ArrowFlightJdbcVarCharVectorAccessorTest.java (hunk 9 above)
Comment (iurysalino) [reply java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/accessor/impl/text/ArrowFlightJdbcVarCharVectorAccessorTest.java:594]: Ok, thanks.

File: java/flight/flight-jdbc-driver/src/main/java/org/apache/arrow/driver/jdbc/accessor/impl/text/ArrowFlightJdbcVarCharVectorAccessor.java (hunk 7 above)
Comment (iurysalino) [reply java/flight/flight-jdbc-driver/src/main/java/org/apache/arrow/driver/jdbc/accessor/impl/text/ArrowFlightJdbcVarCharVectorAccessor.java:110]: fixed

**Review by jcralmeida — CHANGES_REQUESTED** [review]
Resolve the conflicts and then it is approved

File: java/flight/flight-jdbc-driver/src/main/java/org/apache/arrow/driver/jdbc/accessor/impl/text/ArrowFlightJdbcVarCharVectorAccessor.java [hunk 10]
```diff
@@ -100,9 +100,15 @@ public String getString() {
...
-  public boolean getBoolean() {
+  public boolean getBoolean() throws SQLException {
     String value = getString();
-    return value != null && !value.isEmpty() && !value.equals("false") && !value.equals("0");
+    if (value == null || value.equalsIgnoreCase("false") || value.equals("0")) {
+      return false;
+    } else if (value.equalsIgnoreCase("true") || value.equals("1")) {
+      return true;
+    } else {
+      throw new SQLException("It is not possible to convert this value for boolean: " + value);
```
Comment (rafael-telles) [inline java/flight/flight-jdbc-driver/src/main/java/org/apache/arrow/driver/jdbc/accessor/impl/text/ArrowFlightJdbcVarCharVectorAccessor.java:110]:
```suggestion
      throw new SQLException("It is not possible to convert this value to boolean: " + value);
```

**Review by rafael-telles — CHANGES_REQUESTED** [review]

File: java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/accessor/impl/text/ArrowFlightJdbcVarCharVectorAccessorTest.java [hunk 11]
```diff
@@ -593,19 +593,23 @@ public void testShouldGetTimestampReturnValidDateWithCalendar() throws Exception
...
+
   @Test
-  public void testShouldGetBooleanReturnTrueForNonEmpty() throws Exception {
-    assertGetBoolean(new Text("anything"), true);
+  public void testShouldGetBooleanReturnTrueForNonEmpty() {
```
Comment (rafael-telles) [inline java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/accessor/impl/text/ArrowFlightJdbcVarCharVectorAccessorTest.java:606]:
```suggestion
  public void testShouldGetBooleanThrowForInvalidValue() {
```

File: java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/accessor/impl/text/ArrowFlightJdbcVarCharVectorAccessorTest.java [hunk 12]
```diff
@@ -593,19 +593,23 @@ public void testShouldGetTimestampReturnValidDateWithCalendar() throws Exception
...
+    assertGetBooleanForSQLException(new Text("anything"));
   }
 
   @Test
   public void testShouldGetBooleanReturnFalseForEmpty() throws Exception {
```
Comment (rafael-telles) [inline java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/accessor/impl/text/ArrowFlightJdbcVarCharVectorAccessorTest.java:611]:
```suggestion
  public void testShouldGetBooleanThrowForEmpty() throws Exception {
```

**Review by rafael-telles — CHANGES_REQUESTED** [review]

File: java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/accessor/impl/text/ArrowFlightJdbcVarCharVectorAccessorTest.java [hunk 13]
```diff
@@ -593,19 +593,23 @@ public void testShouldGetTimestampReturnValidDateWithCalendar() throws Exception
...
+    when(getter.get(0)).thenReturn(value);
+    collector.checkThrows(SQLException.class, () -> accessor.getBoolean());
+  }
+
   @Test
-  public void testShouldGetBooleanReturnTrueForNonEmpty() throws Exception {
-    assertGetBoolean(new Text("anything"), true);
+  public void testShouldGetBooleanThrowForInvalidValue() {
+    assertGetBooleanForSQLException(new Text("anything"));
   }
```
Comment (jcralmeida) [inline java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/accessor/impl/text/ArrowFlightJdbcVarCharVectorAccessorTest.java:608]: Improve this test name, it is unclear

File: java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/accessor/impl/text/ArrowFlightJdbcVarCharVectorAccessorTest.java [hunk 14]
```diff
@@ -593,19 +593,23 @@ public void testShouldGetTimestampReturnValidDateWithCalendar() throws Exception
...
 
   @Test
-  public void testShouldGetBooleanReturnFalseForEmpty() throws Exception {
-    assertGetBoolean(new Text(""), false);
+  public void testShouldGetBooleanThrowForEmpty() throws Exception {
+    assertGetBooleanForSQLException(new Text(""));
   }
```
Comment (jcralmeida) [inline java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/accessor/impl/text/ArrowFlightJdbcVarCharVectorAccessorTest.java:613]: Improve this test name, it is unclear

**Review by jcralmeida — CHANGES_REQUESTED** [review]

File: java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/accessor/impl/text/ArrowFlightJdbcVarCharVectorAccessorTest.java (hunk 13 above)
Comment (iurysalino) [reply java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/accessor/impl/text/ArrowFlightJdbcVarCharVectorAccessorTest.java]: fixed

File: java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/accessor/impl/text/ArrowFlightJdbcVarCharVectorAccessorTest.java (hunk 14 above)
Comment (iurysalino) [reply java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/accessor/impl/text/ArrowFlightJdbcVarCharVectorAccessorTest.java]: fixed
